File: 97_interleaving_string.js
var isInterleave = function (s1, s2, s3) {
  if (s1.length + s2.length !== s3.length) {
    return false;
  }

  const dp = Array.from({ length: s1.length + 1 }, () =>
    Array(s2.length + 1).fill(false)
  );
  dp[0][0] = true;

  for (let i = 0; i <= s1.length; i++) {
    for (let j = 0; j <= s2.length; j++) {
      if (i > 0 && s1[i - 1] === s3[i + j - 1]) {
        dp[i][j] = dp[i][j] || dp[i - 1][j];
      }
      if (j > 0 && s2[j - 1] === s3[i + j - 1]) {
        dp[i][j] = dp[i][j] || dp[i][j - 1];
      }
    }
  }

  return dp[s1.length][s2.length];
};

// Test cases (Uncomment these if you want to run locally)
// console.log(isInterleave("aabcc", "dbbca", "aadbbcbcac")); // true
// console.log(isInterleave("aabcc", "dbbca", "aadbbbaccc")); // false
// console.log(isInterleave("", "", "")); // true

File: 97_interleaving_string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        # If the length of s3 is not equal to the sum of lengths of s1 and s2, return False
        if len(s1) + len(s2) != len(s3):
            return False
        
        # DP table initialization
        dp = [[False] * (len(s2) + 1) for _ in range(len(s1) + 1)]
        
        # Base case: empty strings
        dp[0][0] = True
        
        # Fill DP table
        for i in range(len(s1) + 1):
            for j in range(len(s2) + 1):
                if i > 0 and s1[i - 1] == s3[i + j - 1]:
                    dp[i][j] = dp[i][j] or dp[i - 1][j]
                if j > 0 and s2[j - 1] == s3[i + j - 1]:
                    dp[i][j] = dp[i][j] or dp[i][j - 1]
        
        # The answer is whether we can form s3 using all of s1 and s2
        return dp[len(s1)][len(s2)]

# Example usage:
solution = Solution()
print(solution.isInterleave("aabcc", "dbbca", "aadbbcbcac"))  # Output: True
print(solution.isInterleave("aabcc", "dbbca", "aadbbbaccc"))  # Output: False
print(solution.isInterleave("", "", ""))  # Output: True
